libclamav_rust/src/codesign.rs

```diff
@@ -0,0 +1,563 @@
+    // now convert the intermediate_cert_path_strs to a Vec<&Path>
+    let intermediate_cert_paths: Vec<PathBuf> = intermediate_cert_path_strs
+        .into_iter()
+        .filter_map(|&path_str| -> Option<PathBuf> {
+            let path_str = if path_str.is_null() {
+                warn!("Intermiediate path string is NULL");
+                return None;
+            } else {
+                #[allow(unused_unsafe)]
+                match unsafe { CStr::from_ptr(path_str) }.to_str() {
+                    Err(e) => {
+                        warn!("Intermediate path string is not valid unicode: {}", e);
+                        return None;
+                    }
+                    Ok(s) => Some(s),
+                }
+            };
+
+            if let Some(path_str) = path_str {
+                match Path::new(path_str).canonicalize() {
+                    Ok(path) => Some(path),
+                    Err(e) => {
+                        warn!(
+                            "Invalid intermediate certificate path: '{}' {}",
+                            path_str, e
+                        );
+                        return None;
+                    }
+                }
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    let signing_key_path_str = validate_str_param!(signing_key_path_str);
+    let signing_key_path = match Path::new(signing_key_path_str).canonicalize() {
+        Ok(p) => p,
+        Err(e) => {
+            return ffi_error!(
+                err = err,
+                Error::SignFailed(format!(
+                    "Invalid signing key path '{}': {}",
+                    signing_key_path_str, e
+                ))
+            );
+        }
+    };
+
+    match sign_file(
+        &target_file_path,
+        &signature_file_path,
+        &signing_cert_path,
+        &intermediate_cert_paths,
+        &signing_key_path,
+    ) {
+        Ok(()) => {
+            debug!("File signed successfully");
+            true
+        }
+        Err(e) => {
+            return ffi_error!(err = err, e);
+        }
+    }
+}
+
+/// Signs a file.
+/// The signature is appended to the signature file.
+pub fn sign_file<P>(
+    target_file_path: &Path,
+    signature_file_path: &Path,
+    signing_cert_path: &Path,
+    intermediate_cert_paths: &[P],
+    signing_key_path: &Path,
+) -> Result<(), Error>
+where
+    P: AsRef<Path>,
+{
+    let signer = Signer::new(signing_key_path, signing_cert_path, intermediate_cert_paths)?;
+
+    let data = std::fs::read(target_file_path)?;
+    let pkcs7 = signer.sign(&data)?;
+
+    // Now convert the pkcs7 to a DigitalSig struct which may be converted to a .sign file signature line.
+    let signature = DigitalSig::Pkcs7(pkcs7);
+    let mut sig_bytes: SigBytes = SigBytes::new();
+    signature.append_sigbytes(&mut sig_bytes)?;
+
+    // If the signature file already exists, open it and append the signature to it.
```

Comment:
Consider adding a parameter to replace the .sign file with a new, or a append it (up to the user).